collection-scripts/gather_istio

```diff
@@ -0,0 +1,214 @@
+  local namespace="${1}"
+  local revision="${2}"
+
+  local istiodName
+  istiodName=$(oc get pod -n "${namespace}" -l "app=istiod,istio.io/rev=${revision}" -o jsonpath="{.items[0].metadata.name}")
+
+  echo "Collecting /debug/syncz from ${istiodName} in namespace ${namespace}"
+
+  local logPath=${BASE_COLLECTION_PATH}/namespaces/${namespace}/${revision}
+  mkdir -p "${logPath}"
+  oc exec "${istiodName}" -n "${namespace}" -c discovery -- /usr/local/bin/pilot-discovery request GET /debug/syncz > "${logPath}/debug-syncz.json" 2>&1
+}
+
+# getEnvoyConfigForPodsInNamespace dumps the envoy config for the specified namespace and
+# control plane revision to a file in the must-gather directory for each pod
+# Arguments:
+#   namespace of the control plane
+#   revision of the control plane
+#   namespace to dump
+# Returns:
+#   nothing
+function getEnvoyConfigForPodsInNamespace() {
+  local controlPlaneNamespace="${1}"
+  local revisionName="${2}"
+  local podNamespace="${3}"
+
+  local istiodName
+  istiodName=$(oc get pod -n "${controlPlaneNamespace}" -l "app=istiod,istio.io/rev=${revisionName}" -o jsonpath="{.items[0].metadata.name}")
+
+  echo "Collecting Envoy config for pods in ${podNamespace} pointing to ${revisionName} revision"
+
+  local pods
+  pods="$(oc get pods -n "${podNamespace}" -o jsonpath='{ .items[*].metadata.name }')"
+  for podName in ${pods}; do
+    if oc get pod -o yaml "${podName}" -n "${podNamespace}" | grep -q proxyv2; then
+      echo "Collecting config_dump and stats for pod ${podName}.${podNamespace}"
+
+      local logPath=${BASE_COLLECTION_PATH}/namespaces/${podNamespace}/pods/${podName}
+      mkdir -p "${logPath}"
+
+      oc exec "${istiodName}" -n "${controlPlaneNamespace}" -c discovery -- bash -c "/usr/local/bin/pilot-discovery request GET /debug/config_dump?proxyID=${podName}.${podNamespace}" > "${logPath}/config_dump_istiod.json" 2>&1
+      oc exec -n "${podNamespace}" "${podName}" -c istio-proxy -- /usr/local/bin/pilot-agent request GET config_dump > "${logPath}/config_dump_proxy.json" 2>&1
+      oc exec -n "${podNamespace}" "${podName}" -c istio-proxy -- /usr/local/bin/pilot-agent request GET stats > "${logPath}/proxy_stats" 2>&1
+    fi
+  done
+}
+
+# Inspect given resource in given namespace (optional)
+# It's using 'oc adm inspect' which will get debug information for given and
+# related resources. Including pod logs.
+# Since OCP 4.16 it honors '--since' and '--since-time' args
+function inspect() {
+  local resource ns
+  resource=$1
+  ns=$2
+
+  if [ -n "$ns" ]; then
+    echo "Inspecting resource ${resource} in namespace ${ns}"
+    # it's here just to make the linter happy (we have to use double quotes arround the variable)
+    if [ -n "${log_collection_args}" ]
+    then
+      oc adm inspect "${log_collection_args}" "--dest-dir=${BASE_COLLECTION_PATH}" "${resource}" -n "${ns}"
+    else
+      oc adm inspect "--dest-dir=${BASE_COLLECTION_PATH}" "${resource}" -n "${ns}"
+    fi
+  else
```

Comment:
Should this be an if statement as it's possible that resource is empty?!?

Should we also check in the first if clause that resources is set? Or possible as a precursor to entering this logic segment?

Reply:
Sure, I'll add an upfront check that $resource is not empty. Currently by inspection we can see that all invocations of this function pass a non-empty first argument, but never hurts to be more defensive.